plutus-core/plutus-ir/Language/PlutusIR/Transform/Inline.hs

```diff
@@ -0,0 +1,245 @@
+basic issues, not do serious optimization. In many cases we've already run the GHC simplifier
+on our input!
+
+We differ from the paper a few ways, mostly leaving things out:
+
+Functionality
+------------
+
+PreInlineUncoditional: we don't do it, since we don't bother using usage information.
+We *could* do it, but it doesn't seem worth it.
+
+CallSiteInline: not worth it.
+
+Inlining recursive bindings: not worth it, complicated.
+
+Context-based inlining: we don't do CallSiteInline, so no point.
+
+Beta reduction: not worth it, but would be easy. We could do the inlining of the argument
+here and also detect dead immediately-applied-lambdas in the dead code pass.
+
+Implementation
+--------------
+
+In-scope set: we don't bother keeping it, since we only ever substitute in things that
+don't have bound variables. This is largely because we don't do PreInlineUnconditional, which
+would inline big things that were only used once, including lambdas etc.
+
+Suspended substitutions for values: we don't do it, since you only need it for
+PreInlineUnconditional
+
+Optimization after substituting in DoneExprs: we can't make different inlining decisions
+contextually, so there's no point doing this.
+-}
+
+
+-- 'SubstRng' in the paper, no 'Susp' case
+-- See Note [Inlining approach and 'Secrets of the GHC Inliner']
+data InlineTerm tyname name uni a = Done (Term tyname name uni a)
+
+newtype TermEnv tyname name uni a = TermEnv { unTermEnv :: UniqueMap TermUnique (InlineTerm tyname name uni a) }
+    deriving newtype (Semigroup, Monoid)
+
+newtype Subst tyname name uni a = Subst { sTermEnv :: TermEnv tyname name uni a }
+    deriving newtype (Semigroup, Monoid)
+
+type ExternalConstraints tyname name uni a term =
+    ( HasUnique name TermUnique
+    , HasUnique tyname TypeUnique
+    , HasConstantIn uni term
+    , PLC.GShow uni
+    , PLC.GEq uni
+    , PLC.DefaultUni PLC.<: uni)
+
+type Inlining tyname name uni a term m =
+    ( MonadState (Subst tyname name uni a) m
+    , MonadReader (Deps.StrictnessMap, DynamicBuiltinNameMeanings term) m
+    , ExternalConstraints tyname name uni a term)
+
+lookupSubst
+    :: (HasUnique name TermUnique)
+    => name
+    -> Subst tyname name uni a
+    -> Maybe (InlineTerm tyname name uni a)
+lookupSubst n (Subst (TermEnv env)) = lookupName n env
+
+extendSubst
+    :: (HasUnique name TermUnique)
+    => name
+    -> InlineTerm tyname name uni a
+    -> Subst tyname name uni a
+    -> Subst tyname name uni a
+extendSubst n clos (Subst (TermEnv env)) = Subst $ TermEnv $ insertByName n clos env
+
+{- Note [Inlining and global uniqueness]
+Inlining relies on global uniqueness (we store things in a unique map), and *does* currently
+preserve it because we don't currently inline anything with binders.
+
+If we do start inlining things with binders in them, we should probably try and preserve it by
+doing something like 'The rapier' section from the paper.
```

Comment:
I think the rapier is needed more for inlining under binders in the absence of global uniqueness rather than inlining things with binders? I mean, if you do inline things with binders, you can just store them in a `Map` wrapped in `Dupable` like we do in the type checker, so that at each lookup site all binders get uniquified.

Reply:
Right, you could do that: essentially just `rename` everything before you substitute it in. The paper argues that this is inefficient, and it's one of the things that the rapier does for you.

I think handling non-globally unique scope is a combination of the rapier and the use of the in-scope set (which we also don't include for that reason).